### ISSUE TITLE
Support creation of a `zbus::Connection` from a `gio::DBusConnection`

### ISSUE THREAD

**Issue by zeenix** [dbus2/zbus#592]
Support creation of a `zbus::Connection` from a [`gio::DBusConnection`](https://docs.rs/gio/latest/gio/struct.DBusConnection.html), which will mean we can finally remove [this thorn](https://gitlab.gnome.org/GNOME/glib/-/issues/2756). However, this will require a bit more. At least, we'll also need to figure out what to do with names registered by the underlying connection. This may or may not work out but it's worth a try.

Requires #590.

**Comment by zeenix** [comment]
Last night I got #590 almost complete and thought more about this. I realized that `gio::DBusConnection` doesn't give you a direct way to receive messages but rather has its own internal ObjectServer, similar to ours. So this will also require us to add mechanism for external ObjectServer impls. We don't necessarily need more public API for that though (at least for this alone), except for a builder/connection constructor that takes a `gio::DBusConnection` instance.

Also while there are[ `send_message*` ](https://docs.rs/gio/latest/gio/struct.DBusConnection.html#method.send_message) for sending, we can't use them for method calls, since there is no way to receive their replies.

Due to this, I'm not sure #590 can help much here. We'll probably need some wrapper instead that `zbus::Connection` can use.

**Comment by REALERvolker1** [comment]
I think this would be a great idea -- I would really like to see this, but I think it would be good to make it available under a feature flag (gtk libraries take literally forever to build). I am making a gtk app right now and it would really help to be able to use only one dbus implementation.

**Comment by zeenix** [comment]
> but I think it would be good to make it available under a feature flag (gtk libraries take literally forever to build)

Of course, that's implied here. 😊 We're not going to depend on gtk though, just gio.

**Comment by zeenix** [comment]
> I am making a gtk app right now and it would really help to be able to use only one dbus implementation.

That is already possible. You don't have to use gio's connection for any of you D-Bus needs at all.

The only "issue" is that `gio::Application` (and `gtk::Application` in turn) takes the application name on the bus so you can't use that with zbus then.

But that's not really a problem, since [it's a good D-Bus practice to append version numbers to your names](https://dbus.freedesktop.org/doc/dbus-api-design.html#api-versioning) anyway. So there doesn't need to be a conflict and in fact there's only a conflict if you don't follow recommended good practices.

**Comment by REALERvolker1** [comment]
Glad to hear this is still planned. I have been trying out `gio::DBusProxy` a bit and zbus is way easier to work with -- more "rusty".

**Comment by zeenix** [comment]
> Glad to hear this is still planned.

I wouldn't call this a plan. I created this issue with the assumption that #590 will enable this [but I was wrong](https://github.com/dbus2/zbus/issues/592#issuecomment-1946719887). Now I'm not sure how we can implement this. I'll probably not be the one doing it anyway.

As I explained above, this isn't really an issue, especially if people would version their service names, like they should.
